libraries/chain/controller.cpp

```diff
@@ -3030,19 +3030,96 @@ struct controller_impl {
+
+            if( bsp->valid_qc->is_strong() && bsp->core.final_on_strong_qc_block_num ) {
+               // We evaluate a block extension qc and advance lib if strong.
+               // This is done before evaluating the block. It is possible the block
+               // will not be valid or forked out. This is safe because the block is
+               // just acting as a carrier of this info. It doesn't matter if the block
+               // is actually valid as it simply is used as a network message for this data.
+               set_if_irreversible_block_num(*bsp->core.final_on_strong_qc_block_num);
+            }
+         }
       }
    }
 
+   // Verify QC claim made by instant_finality_extension in block header extension
+   // and quorum_certificate_extension in block extension are valid.
+   // Called from net-threads. It is thread safe as signed_block is never modified
+   // after creation.
+   void verify_qc_claim( const signed_block_ptr& b, const block_header_state& prev ) {
+      // If block header extension does not have instant_finality_extension,
+      // just return
+      std::optional<block_header_extension> header_ext = b->extract_header_extension(instant_finality_extension::extension_id());
+      if( !header_ext ) {
+         return;
+      }
+
+      // If qc_info is not included, just return
+      const auto& if_ext = std::get<instant_finality_extension>(*header_ext);
+      if( !if_ext.qc_info ) {
+         return;
+      }
+
+      // extract received QC information
+      qc_info_t received_qc_info{ *if_ext.qc_info };
+
+      // if previous block's header extension has the same claim, just return
+      // (previous block had already validated the claim)
+      std::optional<block_header_extension> prev_header_ext = prev.header.extract_header_extension(instant_finality_extension::extension_id());
+      if( prev_header_ext ) {
+         auto prev_if_ext = std::get<instant_finality_extension>(*prev_header_ext);
+         if( prev_if_ext.qc_info && prev_if_ext.qc_info->last_qc_block_num == received_qc_info.last_qc_block_num && prev_if_ext.qc_info->is_last_qc_strong == received_qc_info.is_last_qc_strong ) {
+            return;
+         }
+      }
+
+      // extract QC from block extensions
+      auto block_exts = b->validate_and_extract_extensions();
+      EOS_ASSERT( block_exts.count(quorum_certificate_extension::extension_id()) > 0,
+                  block_validate_exception,
+                  "received block has finality header extension but does not have QC block extension" );
+      const auto& qc_ext = std::get<quorum_certificate_extension>(block_exts. lower_bound(quorum_certificate_extension::extension_id())->second);
+      const auto& received_qc = qc_ext.qc;
+      const auto& valid_qc = qc_ext.qc.qc;
+
+      // Check QC information in header extension and block extension match
+      EOS_ASSERT( received_qc.block_height == received_qc_info.last_qc_block_num,
```

Comment:
I don't quite understand the comment: _This check will not occur if the claim is the same as the last one_. That check was intended to make sure the `block_height` in QC block extension matches `last_qc_block_num` in QC header extension.

Is `block_height` in QC block extension duplicate? I can remove it in a separate issue.

Reply:
If the claim on the current block is the same as the claim in the previous block, this code will not care whether there is a QC in the block extension or not. Here I was just suggesting we should ensure there is no QC provided in that case since it would be redundant.

Reply:
Thanks for the explanation. I understand it now.